### PR TITLE
Remove all vectors of extents

### PR DESCRIPTION
**Pull request by ldorau**
This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/162)

### CODE REVIEW

File: src/vmemcache_heap.c [hunk 1]
```diff
@@ -35,128 +35,157 @@
...
 	return 0;
 }
 
 /*
- * vmcache_heap_get -- get the free heap entry from the heap
+ * vmcache_get_heap_entry -- get the free entry from the heap
  */
 static inline int
-vmcache_heap_get(struct heap *heap, struct heap_entry *he)
+vmcache_get_heap_entry(struct heap *heap, struct heap_entry *he)
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: this isn't 'get', it's a 'pop' or 'remove' operation.

File: src/vmemcache_heap.c [hunk 2]
```diff
@@ -198,59 +226,105 @@ vmcache_heap_destroy(struct heap *heap)
...
-			to_allocate = 0;
-			break;
+		/* allocated size without headers */
+		size_t allocated_size = he.size - HFER_SIZE;
+		allocated += allocated_size;
+
+		if (allocated_size > to_allocate &&
+		    allocated_size - to_allocate >= extent_size - HFER_SIZE &&
+		    first_extent->size == extent_size - HFER_SIZE) {
+			vmcache_free_first_extent(heap, first_extent);
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: Does it matter that the extent is first? Shouldn't this be "free any extent that has size == extent_size" ?

File: src/vmemcache_heap.c [hunk 3]
```diff
@@ -198,59 +226,105 @@ vmcache_heap_destroy(struct heap *heap)
...
-		if (vmcache_heap_vec_insert(vec, &he)) {
+		if (vmcache_insert_heap_entry(heap, &he, last_extent,
+							IS_ALLOCATED)) {
 			util_mutex_unlock(&heap->lock);
 			return -1;
 		}
 
-		/* size without headers */
-		allocated += he.size;
+		if (first_extent->ptr == NULL && to_allocate == size) {
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: `to_allocate == size` will always be true if `first_extent->ptr == NULL` is true.

File: src/vmemcache_heap.c [hunk 4]
```diff
@@ -261,47 +335,71 @@ vmcache_alloc(struct heap *heap, size_t size, struct extent_vec *vec)
...
+	if (header->next) {
+		struct header *header_next = HEADER(header->next);
+		ASSERTeq(header_next->prev, ext->ptr);
+		header_next->prev = header->prev;
+	}
+
+	if (heap->last_extent == ext->ptr)
+		heap->last_extent = header->prev;
+
+	__sync_fetch_and_sub(&heap->entries, 1);
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: util_

File: src/vmemcache_heap.c [hunk 5]
```diff
@@ -35,128 +35,157 @@
...
+	if (*last_extent) {
+		struct header *last_header = HEADER(*last_extent);
+		ASSERTeq(last_header->next, NULL);
+		last_header->next = new_extent;
+	}
+
+	*last_extent = new_extent;
+
+	if (!is_allocated) {
+		__sync_fetch_and_add(&heap->entries, 1);
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: util_

File: src/vmemcache_heap.h [hunk 6]
```diff
@@ -49,24 +49,45 @@ extern "C" {
...
+void vmcache_free(struct heap *heap, void *last_extent);
 
 stat_t vmcache_get_heap_used_size(struct heap *heap);
 stat_t vmcache_get_heap_entries_count(struct heap *heap);
 
+void *vmcache_extent_get_prev(void *ptr);
+size_t vmcache_extent_get_size(void *ptr);
+
+/* unsafe variant - the headers of extents cannot be modified */
+#define EXTENTS_FOREACH(ext, extents) \
```
Comment (pbalcer) [inline src/vmemcache_heap.h]: surround macro arguments in () when possible

File: src/vmemcache_heap.h [hunk 7]
```diff
@@ -49,24 +49,45 @@ extern "C" {
...
+	for (ext.ptr = extents, \
+		ext.size = ext.ptr ? vmcache_extent_get_size(ext.ptr) : 0; \
```
Comment (pbalcer) [inline src/vmemcache_heap.h]: you could simplify this by allowing NULL in _get_size and _get_prev. No more ternary operators ;)

File: src/vmemcache_heap.c [hunk 8]
```diff
@@ -198,59 +226,105 @@ vmcache_heap_destroy(struct heap *heap)
...
- * otherwise -1 is returned.
+ * It returns the number of allocated bytes if successful, otherwise -1.
+ * The last extent of doubly-linked list of allocated extents is returned
+ * in 'last_extent'.
+ * 'first_extent' has to be zeroed in the beginning of a new allocation
+ * (e.g. when *last_extent == NULL).
  */
 ssize_t
-vmcache_alloc(struct heap *heap, size_t size, struct extent_vec *vec)
+vmcache_alloc(struct heap *heap, size_t size, void **last_extent,
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: I think this patch would be more readable with the extent lists beginning at the front ;)

File: src/vmemcache_heap.c (hunk 1 above)
Comment (ldorau) [reply src/vmemcache_heap.c]: Done

File: src/vmemcache_heap.c (hunk 2 above)
Comment (ldorau) [reply src/vmemcache_heap.c]: Right, done.

File: src/vmemcache_heap.c (hunk 4 above)
Comment (ldorau) [reply src/vmemcache_heap.c]: Done

File: src/vmemcache_heap.c (hunk 5 above)
Comment (ldorau) [reply src/vmemcache_heap.c]: Done

File: src/vmemcache_heap.c (hunk 3 above)
Comment (ldorau) [reply src/vmemcache_heap.c]: Changed

File: src/vmemcache_heap.h (hunk 6 above)
Comment (ldorau) [reply src/vmemcache_heap.h]: Done

File: src/vmemcache_heap.h (hunk 7 above)
Comment (ldorau) [reply src/vmemcache_heap.h]: I thought this way is faster. Changed.

File: src/vmemcache_heap.c (hunk 8 above)
Comment (ldorau) [reply src/vmemcache_heap.c]: Why? I am not sure about it.

File: src/vmemcache_heap.c [hunk 9]
```diff
@@ -198,110 +230,186 @@ vmcache_heap_destroy(struct heap *heap)
...
-	VEC_DELETE(&heap->entries);
 	util_mutex_destroy(&heap->lock);
 	Free(heap);
 }
 
+/*
+ * vmcache_free_small_extent -- (internal) free the smallest extent
+ */
+static int
+vmcache_free_small_extent(struct heap *heap, void *small_extent)
```
Comment (pbalcer) [inline src/vmemcache_heap.c]: `vmemcache_free_extent`.

Anyway, we have three different functions to remove an extent from a list with only slight variations. Is there a way to reduce the repetitions?

**Comment by ldorau** [comment]
@marcinslusarz next <-> prev && first <-> last - done

**Comment by ldorau** [comment]
@marcinslusarz : type safety - done

**Comment by ldorau** [comment]
@pbalcer please resolve your discussions

**Comment by ldorau** [comment]
@marcinslusarz  the `VOID_PTR` macro removed

**Pull request merged**